### ISSUE TITLE
Model Load from a brand new workflow

### ISSUE THREAD

**Issue by albertodema** [salesforce/TransmogrifAI#75]
**Problem**
Actually in order to score a model that is a result of an automl selection I have first to train the model, save it to disk, load it and score against the scoring dataset and all of this it is performed in the same method 

**Solution**
Once a model is saved, it should be possible to load it without the need of the original workflow that created the model itself.

**Alternatives**
Actually I do not find alternatives.

**Comment by tovbinm** [comment]
Currently the limitation is the lambda transformers, which do not allow us to fully use reflection when loading model stages. One potential solution I have is to replace `feature.map(...)` shortcuts (which use lambda transformers underneath) with a macro that would actually generate a concrete class transformer for us. That way loading a model would not require the workflow, as we would serialize and reflect by class name.

**Comment by allwefantasy** [comment]
Model Load from a brand new workflow is really important. Now I am trying to fix this with your solution mentioned before.  We modify OpPiplineStageReader 👍 

```scala
val stage = if (isModel) loadModel(className, metadataJson) else loadTransformer(className, metadataJson)
```

Make sure all stage revover from metadataJson.

**Comment by tovbinm** [comment]
@allwefantasy there is no easy fix that I can think of. you need the exact same workflow (with the exact same stages in it) to load a trained model. We can continue this discussion on gitter - https://gitter.im/salesforce-TransmogrifAI-Issues/Issues

**Comment by allwefantasy** [comment]
@tovbinm  Yes.....I have failed to fix this since there are to many things need to do. Does the community have the plan to fix this?

**Comment by tovbinm** [comment]
@allwefantasy Nobody participated really yet. Right now My team and I are trying to think of a better solution on how to handle loading models. I will share the design document once we have something.

As of right now - as long as you use the same workflow you should be fine. Can you now guarantee that in your case?

**Comment by albertodema** [comment]
Hi can we somehow use this for model scoring without the need to recreate the same workflow?
https://github.com/salesforce/TransmogrifAI/pull/41

**Comment by tovbinm** [comment]
no, the requirement stands. loading model is only possible using the same workflow. I think it would be beneficial if you could share a bit more on what are you trying to do, so I could help you.

**Comment by albertodema** [comment]
As you know I am working on this: http://mlforall.azurewebsites.net/
My need is the following : once the model is trained with an on-demand spark cluster of containers (today just 1 :-) ) and using all the automl capabilities of transmogrifai combined with spark data type detection (so no case class or features built by hand plus the sanity checks) .
I would like to perform scoring quickly without having to re-do the auto-ml steps of recalculating the features from the entire dataset as it was done in the training and ideally using a serverless framework (like azure-function or amazon lambda etc..) that is able to work with jvm.
So in my case this avoidance of rebuilding the same workflow it's very important because for scoring it would take sometimes even 30 sec or sometimes 5 minutes to just re-do the auto-feature part and of course this is not acceptable for quick scoring where you want to do, among other things, for example quick what if scenarios.

**Comment by tovbinm** [comment]
Rebuilding the workflow is fast, but retraining the model is not. Do you happen to have the code available somewhere? I would be glad to have a look.

**Comment by albertodema** [comment]
It is quite straightforward  and it is an example of scoring on all the training dataset .
Once the data is inside the dataframe passengersData and with some queries on the target variable I detect if I have to load a  binary or regression or multiclassmodel:

if (detected=="Binary') {
      val (saleprice, features) = FeatureBuilder.fromDataFrame[RealNN](passengersData, response = targetColumn)
      val featureVector = features.toSeq.autoTransform()

      val checkedFeatures = saleprice.sanityCheck(featureVector, checkSample = 1.0, removeBadFeatures = true)
      val (pred, rawPrediction, prob) = BinaryClassificationModelSelector().setInput(saleprice, checkedFeatures).getOutput()
      val wf = new OpWorkflow()
      val model = wf.setResultFeatures(pred,prob)
        .loadModel("/models/" + uniqueId + "/binmodel/")
        .setInputDataset(passengersData)
      val dataframe = model.score(None,true)
  


    }
    else if (detected=="Multiclass") {

      val (saleprice, features) = FeatureBuilder.fromDataFrame[Text](passengersData, response = targetColumn)
      val featureVector = features.toSeq.autoTransform()
      val (pred, rawPrediction, prob) = MultiClassificationModelSelector().setInput(saleprice.indexed(), featureVector).getOutput()
      val wf = new OpWorkflow()
      val model = wf.setResultFeatures(pred,prob).loadModel("models/" + uniqueId + "/multicmodel/").setInputDataset(passengersData)
      val dataframe = model.score(None,true)
    }
    else { //So it's a regression
      val (saleprice, features) = FeatureBuilder.fromDataFrame[RealNN](passengersData, response = targetColumn)
      val featureVector = features.toSeq.autoTransform()
      val checkedFeatures = saleprice.sanityCheck(featureVector, checkSample = 1.0, removeBadFeatures = true)
      val pred = RegressionModelSelector().setInput(saleprice, checkedFeatures).getOutput()
      val wf = new OpWorkflow()
      val model = wf.setResultFeatures(pred).loadModel("/models/" + uniqueId + "/regmodel/").setInputDataset(passengersData)
      val dataframe = model.score(None,true)
      


    }

Depending on the amount of data especially the sanity checks can take minutes....

**Comment by tovbinm** [comment]
Since you are not training a model here, it should not consume significant time. How did you measure that sanity check takes time?

**Comment by albertodema** [comment]
i will collect some metrics, with public datasets and post them here so you can reproduce.

**Comment by tovbinm** [comment]
Thank you for the info. This is very helpful.

From what I can see from the logs majority of the time is spend on loading the model and not constructing the workflow (which as you can guess from the logs uses Spark parquet reader).
Then the passenger csv file reading is done (last 4 seconds).

As for the model loading part, I would recommend to load the model and keep it in process memory, or perhaps as you pointed out before - use the [local module](https://github.com/salesforce/TransmogrifAI/tree/master/local) that we just released.
So once the model is trained, you would prepare the scoring function right away and keep it in memory of you process keyed by model id:
```
import com.salesforce.op.local._
val scoreFn = model.scoreFunction
```
Then on scoring, you should be able to call it:
```
val passenger = Map("age" -> 18, "name" -> "Peter")
val score: Map[String, Any] = scoreFn(passenger)
```

**Comment by albertodema** [comment]
Thanks for you suggestions @tovbinm .
However I cannot apply those because, as I said, I want to perform this scoring with a serverless framework (azure functions/aws lambda) that of course is a "stateless" app that is triggered and loaded only on request for each specific request. So I cannot put this model in memory, because after each call the entire jvm process goes down and I pay only for the execution time.

**Comment by tovbinm** [comment]
Hmm, how can one implement a state in lambda infrastructure like this? You might need to serialize the scoring function to some model cache, retrieve it from the by model id, deserialize and run it.

You might want to serialize the scoring function using [`ObjectOutputStream`](https://docs.oracle.com/javase/8/docs/api/java/io/ObjectOutputStream.html), but please consider the security implications of such approach.

In general I think TransmogrifAI (an Spark as well) would benefit from functionality of exporting models in a more portable format such as [mleap](https://github.com/combust/mleap) or [PFA](http://dmg.org/pfa/), that would perhaps allow you faster model loading times (which I am not sure if it's the case).

**Comment by albertodema** [comment]
Here an example of what i say using ML.NET:
http://luisquintanilla.me/2018/08/21/serverless-machine-learning-mlnet-azure-functions/
I also believe that if we can have models in PFA or mleap to be loaded quickly without spark context, it would be much better for scoring.

**Comment by tovbinm** [comment]
Workflow independent model loading is allowed once this PR is merged - #274

**Comment by tovbinm** [comment]
:champagne: :fireworks:
